### PR TITLE
Support for in-band-mgmt via management VRF

### PR DESCRIPTION
**Pull request by venkatmahalingam**
Signed-off-by: Venkatesan Mahalingam <venkatesan_mahalinga@dell.com>
Please refer [HLD](https://github.com/Azure/SONiC/pull/638) and [swss-common](https://github.com/Azure/sonic-swss-common/pull/479) pull requests for more information.

**What I did**

Support for in-band-mgmt via management VRF

**Why I did it**
Other than eth0, mgmt VRF should have in-band L3 interfaces as well.

**How I verified it**
Verified the functionality using sonic-vs testcases

**Details if related**

Please refer [old PR](https://github.com/Azure/sonic-swss/pull/1707) for old comments,

### CODE REVIEW

File: orchagent/intfsorch.cpp [hunk 1]
```diff
@@ -406,6 +423,9 @@ bool IntfsOrch::setIntf(const string& alias, sai_object_id_t vrf_id, const IpPre
             intfs_entry.ref_count = 0;
             intfs_entry.proxy_arp = false;
             intfs_entry.vrf_id = vrf_id;
+            if (!vrf_name.empty()) {
```
Comment (prsunny) [inline orchagent/intfsorch.cpp:426]: `{` -> nextline

File: orchagent/intfsorch.cpp [hunk 2]
```diff
@@ -406,6 +423,9 @@ bool IntfsOrch::setIntf(const string& alias, sai_object_id_t vrf_id, const IpPre
...
+                intfs_entry.vrf_name = vrf_name;
```
Comment (prsunny) [inline orchagent/intfsorch.cpp:427]: You don't have to change the API signature - Use `getVRFname `function. May be refer the comment in function `isInbandIntfInMgmtVrf`

File: orchagent/intfsorch.h [hunk 3]
```diff
@@ -52,7 +54,8 @@ class IntfsOrch : public Orch
     void addRifToFlexCounter(const string&, const string&, const string&);
     void removeRifFromFlexCounter(const string&, const string&);
 
-    bool setIntf(const string& alias, sai_object_id_t vrf_id = gVirtualRouterId, const IpPrefix *ip_prefix = nullptr, const bool adminUp = true, const uint32_t mtu = 0);
+    bool setIntf(const string& alias, const string& vrf_name, sai_object_id_t vrf_id = gVirtualRouterId, const IpPrefix *ip_prefix = nullptr,
```
Comment (prsunny) [inline orchagent/intfsorch.h:57]: As mentioned above - use `getVRFname`

File: orchagent/intfsorch.cpp [hunk 4]
```diff
@@ -159,6 +160,21 @@ string IntfsOrch::getRouterIntfsAlias(const IpAddress &ip, const string &vrf_nam
     return string();
 }
 
+bool IntfsOrch::isInbandIntfInMgmtVrf(const string& alias)
+{
+    if (m_syncdIntfses.find(alias) == m_syncdIntfses.end())
```
Comment (prsunny) [inline orchagent/intfsorch.cpp:165]: Please use `{`

File: orchagent/neighorch.cpp [hunk 5]
```diff
@@ -551,7 +551,8 @@ void NeighOrch::doTask(Consumer &consumer)
 
         string alias = key.substr(0, found);
 
-        if (alias == "eth0" || alias == "lo" || alias == "docker0")
+        if (alias == "eth0" || alias == "lo" || alias == "docker0"
+            || ((op == SET_COMMAND) && m_intfsOrch->isInbandIntfInMgmtVrf(alias)))
```
Comment (prsunny) [inline orchagent/neighorch.cpp:555]: How about any routes pointing to inband mgmt vrf?

File: tests/test_inband_intf_mgmt_vrf.py [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+            time.sleep(1)
+
+    def test_MgmtVrf(self, dvs, testlog):
+        self.setup_db(dvs)
+
+        vrf_oid = self.add_mgmt_vrf()
+        self.del_mgmt_vrf()
+
+    @pytest.mark.parametrize('intf_name', ['Ethernet4', 'Vlan100', 'PortChannel5', 'Loopback1'])
+    def test_InbandIntf(self, intf_name, dvs, testlog):
```
Comment (prsunny) [inline tests/test_inband_intf_mgmt_vrf.py:108]: Please add a test for neighbor entries so it will skip installing in ASIC_DB

File: orchagent/intfsorch.cpp [hunk 7]
```diff
@@ -159,6 +160,21 @@ string IntfsOrch::getRouterIntfsAlias(const IpAddress &ip, const string &vrf_nam
...
+        return false;
+
+    string vrf_name = "";
+    vrf_name = m_syncdIntfses[alias].vrf_name;
```
Comment (prsunny) [inline orchagent/intfsorch.cpp:169]: I would suggest, a simple approach would be to use the `getVRFname `from `vrf_id `here. So you dont have to add this vrf_name to the structure.

File: orchagent/intfsorch.cpp (hunk 4 above)
Comment (venkatmahalingam) [reply orchagent/intfsorch.cpp:165]: done

File: orchagent/neighorch.cpp (hunk 5 above)
Comment (venkatmahalingam) [reply orchagent/neighorch.cpp:555]: Those routes would be ignored in fpmsyncd.

File: orchagent/intfsorch.cpp (hunk 7 above)
Comment (venkatmahalingam) [reply orchagent/intfsorch.cpp]: done

File: orchagent/intfsorch.h (hunk 3 above)
Comment (venkatmahalingam) [reply orchagent/intfsorch.h]: ok

File: orchagent/intfsorch.cpp (hunk 2 above)
Comment (venkatmahalingam) [reply orchagent/intfsorch.cpp]: sure

File: orchagent/intfsorch.cpp (hunk 1 above)
Comment (venkatmahalingam) [reply orchagent/intfsorch.cpp]: done

File: cfgmgr/vrfmgr.cpp [hunk 8]
```diff
@@ -235,8 +308,19 @@ void VrfMgr::doTask(Consumer &consumer)
                         it = consumer.m_toSync.erase(it);
                         continue;
                     }
+                    vector<FieldValueTuple> vrfFvVector;
+                    // Incase of MGMT_VRF_CONFIG table event, the fields mgmtVrfEnabled, 
+                    // in_band_mgmt_enabled should be ignored.
+                    if (consumer.getTableName() != CFG_MGMT_VRF_CONFIG_TABLE_NAME)
```
Comment (prsunny) [inline cfgmgr/vrfmgr.cpp:314]: Is this really required. I think it should be ok to leave the fields populated.

File: cfgmgr/vrfmgr.cpp [hunk 9]
```diff
@@ -207,6 +225,60 @@ void VrfMgr::doTask(Consumer &consumer)
...
+                {
+                    if (fvField(i) == "mgmtVrfEnabled")
+                    {
+                        if (fvValue(i) == "true")
+                        {
+                            mgmt_vrf_enabled = true;
+                        }
+                        else
+                        {
+                            op = DEL_COMMAND;
```
Comment (prsunny) [inline cfgmgr/vrfmgr.cpp:246]: why is this else case required to set `op`. You are already doing this in line 266, right? Please remove both the else cases.

Comment (venkatmahalingam) [reply cfgmgr/vrfmgr.cpp:246]: ok

File: cfgmgr/vrfmgr.cpp (hunk 8 above)
Comment (venkatmahalingam) [reply cfgmgr/vrfmgr.cpp:314]: ok, we can populate those fields but it won't be of any use for now.

File: cfgmgr/vrfmgr.cpp [hunk 10]
```diff
@@ -235,8 +299,10 @@ void VrfMgr::doTask(Consumer &consumer)
                         it = consumer.m_toSync.erase(it);
                         continue;
                     }
+                    vector<FieldValueTuple> vrfFvVector;
```
Comment (prsunny) [inline cfgmgr/vrfmgr.cpp:302]: Please revert this change.

File: tests/test_inband_intf_mgmt_vrf.py [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+        intf_keys = tbl.getKeys()
+        status, fvs = tbl.get(intf_name)
+        assert status == True
+        for fv in fvs:
+            if fv[0] == 'vrf_name':
+                assert fv[1] == MGMT_VRF_NAME
+
+        if not intf_name.startswith('Loopback'):
+            # check ASIC router interface database
+            tbl = swsscommon.Table(self.asic_db, 'ASIC_STATE:SAI_OBJECT_TYPE_ROUTER_INTERFACE')
```
Comment (prsunny) [inline tests/test_inband_intf_mgmt_vrf.py:125]: for checking ASIC db, could you please use https://github.com/Azure/sonic-swss/blob/master/tests/test_mux.py#L90
Using sleep may be flaky

**Comment by prsunny** [comment]
@venkatmahalingam, can you take a look at the VS test [failure](https://dev.azure.com/mssonic/build/_build/results?buildId=19533&view=logs&j=3f6395b2-1619-5ebe-f305-2aedcf353cb5&t=35460df8-21be-5313-3ee4-50900c6db765)?

**Comment by venkatmahalingam** [comment]
> @venkatmahalingam, can you take a look at the VS test [failure](https://dev.azure.com/mssonic/build/_build/results?buildId=19533&view=logs&j=3f6395b2-1619-5ebe-f305-2aedcf353cb5&t=35460df8-21be-5313-3ee4-50900c6db765)?

ok

**Comment by prsunny** [comment]
Is the latest change causing VS test failures?

**Comment by venkatmahalingam** [comment]
> Is the latest change causing VS test failures?

yes, fixed all the failures, please review again and merge the PR.

File: cfgmgr/vrfmgr.cpp [hunk 12]
```diff
@@ -235,9 +299,8 @@ void VrfMgr::doTask(Consumer &consumer)
                         it = consumer.m_toSync.erase(it);
                         continue;
                     }
-
+
```
Comment (prsunny) [inline cfgmgr/vrfmgr.cpp:302]: Extra tabspace added

File: tests/test_inband_intf_mgmt_vrf.py [hunk 13]
```diff
@@ -0,0 +1,182 @@
...
+
+        self.remove_inband_intf(intf_name)
+        time.sleep(1)
+        # check application database
+        tbl = swsscommon.Table(self.appl_db, 'INTF_TABLE')
+        intf_keys = tbl.getKeys()
+        assert len(intf_keys) == 0
+
+        if not intf_name.startswith('Loopback'):
+            time.sleep(5)
```
Comment (prsunny) [inline tests/test_inband_intf_mgmt_vrf.py:168]: can we change the sleep to wait for keys?

**Review by prsunny — APPROVED** [review]

File: tests/test_inband_intf_mgmt_vrf.py [hunk 14]
```diff
@@ -0,0 +1,169 @@
+import time
```
Comment (prsunny) [inline tests/test_inband_intf_mgmt_vrf.py:1]: Please delete the binary file tests/.test_interface.py.swp

**Review by prsunny — APPROVED** [review]

**Comment by prsunny** [comment]
Please add sonic-mgmt tests for covering this. I'll add a tracking issue.
